File: code/code15.py
class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        n = len(nums)
        ans = []
        nums.sort()
        for i in range(n):
            if i > 0 and nums[i] == nums[i-1]:
                continue
            r = n - 1
            target = -nums[i]
            for l in range(i+1,n):
                if l > i + 1 and nums[l] == nums[l-1]:
                    continue
                while l < r and nums[l] + nums[r] > target:
                    r -= 1
                if l >= r :
                    break
                if  nums[l] + nums[r] == target:
                    ans.append([-target,nums[l],nums[r]])
        return ans
if __name__ == "__main__":
    solution = Solution()
    nums = [0,0,0]
    print(solution.threeSum(nums))

File: code/code15.cpp
#include <bits/stdc++.h>
using namespace std;
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        int n = nums.size();
        vector<vector<int>> ans;
        sort(nums.begin(), nums.end());
        if (n <= 2) {
            return ans;
        }
        for (auto i = 0;i < n;i++) {
            if (i > 0 && nums[i-1] == nums[i]) {
                continue;
            }
            int t = n - 1;
            int tag = -nums[i];
            for (auto j = i+1;j < n;j++) {
                if (j > i + 1 && nums[j-1] == nums[j]) {
                    continue;
                }
                while (j < t&&nums[j] + nums[t] > tag) {
                    --t;
                }
                if (t == j) {
                    break;
                }
                if (nums[j] + nums[t] == tag) {
                    ans.push_back({nums[i],nums[j],nums[t]});
                }
            }
        }
        return ans;
    }
};